#!/usr/bin/env nextflow

/*
####################################################################################################
################################## Module Kallisto_alignment #######################################
####################################################################################################
*/

nextflow.enable.dsl = 2

/*

################################## Module Kallisto_alignment #######################################

    This module contains the process to execute the Kallisto pseudoaligner that makes a quantification of the transcripts from the fastq files, this process takes as input the subdirectory generated by the CreateDir process (Module_OrderFiles_FastQC_rnaSeq_workflow.nf) and returns the directory where all the results are generated.

*/

if (params.help) {
	log.info ''
	exit 0
}    

process Kallisto_alignment_module {
	
conda ("/home/.conda/envs/ipx-kallisto")

	errorStrategy 'ignore'
	
	input:
	val directory	
	
	output:
	stdout
	
	script:
	"""
	R1=\$(ls \$(echo $directory)/*_R1_*.fastq.gz)
	R2=\$(ls \$(echo $directory)/*_R2_*.fastq.gz)
	path=\$(basename \$(echo \${R1}) | awk -F "_R1" '{print \$1}')
	
	kallisto quant -i ${params.idx_kallisto} -t 15 -b 100 -o \$(echo $directory)/Kallisto_results \${R1} \${R2} > /dev/null 2>&1
	
	mv \$(echo $directory)/Kallisto_results/abundance.tsv \$(echo $directory)/Kallisto_results/\${path}_abundance.tsv
	"""
}
